janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/AdjacentVertexHasUniquePropertyOptimizerStrategy.java

```diff
@@ -0,0 +1,146 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.tinkerpop.optimize;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Iterables;
+import org.apache.tinkerpop.gremlin.process.traversal.*;
+import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.DefaultGraphTraversal;
+import org.apache.tinkerpop.gremlin.process.traversal.step.filter.FilterStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.filter.HasStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.map.GraphStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.map.IdStep;
+import org.apache.tinkerpop.gremlin.process.traversal.step.util.HasContainer;
+import org.apache.tinkerpop.gremlin.process.traversal.util.TraversalHelper;
+import org.apache.tinkerpop.gremlin.structure.Edge;
+import org.apache.tinkerpop.gremlin.structure.Vertex;
+import org.janusgraph.core.Cardinality;
+import org.janusgraph.core.JanusGraphElement;
+import org.janusgraph.core.RelationType;
+import org.janusgraph.graphdb.internal.InternalRelationType;
+import org.janusgraph.graphdb.query.JanusGraphPredicateUtils;
+import org.janusgraph.graphdb.query.QueryUtil;
+import org.janusgraph.graphdb.query.condition.*;
+import org.janusgraph.graphdb.transaction.StandardJanusGraphTx;
+import org.janusgraph.graphdb.types.CompositeIndexType;
+import org.janusgraph.graphdb.types.IndexType;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.janusgraph.graphdb.types.system.ImplicitKey.ADJACENT_ID;
+
+/**
+ * @author Florian Grieskamp (Florian.Grieskamp@gdata.de)
+ */
+public class AdjacentVertexHasUniquePropertyOptimizerStrategy
+    extends AdjacentVertexOptimizerStrategy<HasStep<?>> {
+
+    private static final AdjacentVertexHasUniquePropertyOptimizerStrategy INSTANCE =
+        new AdjacentVertexHasUniquePropertyOptimizerStrategy();
+
+    private AdjacentVertexHasUniquePropertyOptimizerStrategy() {}
+
+    public static AdjacentVertexHasUniquePropertyOptimizerStrategy instance() { return INSTANCE; }
+
+    @Override
+    public void apply(final Traversal.Admin<?, ?> traversal) {
+        TraversalHelper.getStepsOfClass(HasStep.class, traversal)
+            .forEach(this::optimizeStep);
+    }
+
+    /**
+     * Determines whether this HasStep can be answered by a unique index and thus, returns either 0 or 1 match
+     */
+    @Override
+    protected boolean isValidStep(HasStep<?> step) {
+        StandardJanusGraphTx tx;
+
+        try {
+            tx = (StandardJanusGraphTx) JanusGraphTraversalUtil.getTx(step.getTraversal());
+        } catch (IllegalArgumentException iaex) {
+            // The traversal needs to be bound to a graph in order to get information about the indexes
```

Comment:
When would IllegalArgumentException happen? I don't quite understand the logic here.

Reply:
It works perfectly fine without the `try...catch` block except for a single TinkerPop test class which somehow creates traversals which are not bound to a graph. I didn't try to find out how or why these traversals are created that way. But as the `getTx(...)` method explicitly checks if the traversal is bound to a graph, it looks to me like a valid case to have an unbound traversal.
My first plan was to avoid exception handling but then I would have to duplicate most of the code from `getTx(...)`. That's why I went with `try...catch`.

Reply:
On second thought: I can try to extract only the line `traversal.getGraph().isPresent()` and see if that's already sufficient. Just saw that this is how other optimizers handle such cases.

Reply:
Confirmed that it works like I described in the comment above.